server/src/main/java/org/opensearch/index/remote/RemoteIndexPathUploader.java

```diff
@@ -0,0 +1,260 @@
+import org.opensearch.repositories.blobstore.ChecksumBlobStoreFormat;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static org.opensearch.gateway.remote.RemoteClusterStateService.INDEX_METADATA_UPLOAD_TIMEOUT_SETTING;
+import static org.opensearch.index.remote.RemoteIndexPath.SEGMENT_PATH;
+import static org.opensearch.index.remote.RemoteIndexPath.TRANSLOG_PATH;
+import static org.opensearch.node.remotestore.RemoteStoreNodeAttribute.isRemoteDataAttributePresent;
+import static org.opensearch.node.remotestore.RemoteStoreNodeAttribute.isRemoteStoreClusterStateEnabled;
+
+/**
+ * Uploads the remote store path for all possible combinations of {@link org.opensearch.index.remote.RemoteStoreEnums.DataCategory}
+ * and {@link org.opensearch.index.remote.RemoteStoreEnums.DataType} for each shard of an index.
+ */
+public class RemoteIndexPathUploader implements IndexMetadataUploadInterceptor {
+
+    public static final ChecksumBlobStoreFormat<RemoteIndexPath> REMOTE_INDEX_PATH_FORMAT = new ChecksumBlobStoreFormat<>(
+        "remote-index-path",
+        RemoteIndexPath.FILE_NAME_FORMAT,
+        RemoteIndexPath::fromXContent
+    );
+
+    private static final String TIMEOUT_EXCEPTION_MSG = "Timed out waiting while uploading remote index path file for indexes=%s";
+    private static final String UPLOAD_EXCEPTION_MSG = "Exception occurred while uploading remote index paths for indexes=%s";
+
+    private static final Logger logger = LogManager.getLogger(RemoteIndexPathUploader.class);
+
+    private final Settings settings;
+    private final boolean isRemoteDataAttributePresent;
+    private final boolean isTranslogSegmentRepoSame;
+    private final Supplier<RepositoriesService> repositoriesService;
+    private volatile TimeValue indexMetadataUploadTimeout;
+
+    private BlobStoreRepository translogRepository;
+    private BlobStoreRepository segmentRepository;
+
+    public RemoteIndexPathUploader(Settings settings, Supplier<RepositoriesService> repositoriesService, ClusterSettings clusterSettings) {
+        this.settings = settings;
+        this.repositoriesService = repositoriesService;
+        isRemoteDataAttributePresent = isRemoteDataAttributePresent(settings);
+        // If the remote data attributes are not present, then there is no effect of translog and segment being same or different or null.
+        isTranslogSegmentRepoSame = isTranslogSegmentRepoSame();
+        indexMetadataUploadTimeout = clusterSettings.get(INDEX_METADATA_UPLOAD_TIMEOUT_SETTING);
+        clusterSettings.addSettingsUpdateConsumer(INDEX_METADATA_UPLOAD_TIMEOUT_SETTING, this::setIndexMetadataUploadTimeout);
```

Comment:
Raising a separate PR for moving the remote cluster state related settings to a different class so that there can be single source of truth and only one consumer for updating the settings.